### PR TITLE
Created a fully working post todo route

### PR DESCRIPTION
**Pull request by vinay-jtc**
## Description
Created an endpoint to add a new TODO item.

URI: POST /todos
Request body: { title: … }
Response HTTP Status: 201
Response body: { id: …, title: … }

## Database schema changes
New model Todo added.
Schema:
{
title: { type: String, required: true }
},
{
collection: 'todos',
timestamps: {
createdAt: 'createdAt',
updatedAt: 'updatedAt',
},
}

### Automated test cases added

1. Should create a new todo if valid string is passed as "title" in body, it should return 201
2. Should give 400 if empty string is passed as "title" in body
### Manual test cases run

Test cases run:
1) URI: POST /todos
Request body: { "title" : "checking..." }
Response HTTP Status: 201
Response body: { 
"id": "hggfeh63hj7c2he873cu",
 "title": "checking..."
}

2) URI: POST /todos
Request body: { "title" : "" }
Response HTTP Status: 400
Response body: {
    "message": "Validation failed. Please modify the request and try again.",
    "failures": [
            {
                "field": "title",
                "message": "Please specify the valid title"
            }
    ]
}

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,7 +6,8 @@ In order to run the project locally, you need few keys in order for 3rd party se
 
 1. Install dependencies: `npm i`
 2. Reach out to your team member for `local.json` (which has sensitive keys) and drop it into `config` folder
-3. Start the server: `npm run serve`
+3. Start the server: `gulp serve`
```
Comment (coder1033) [inline README.md:9]: `npm run serve` this cmd is already running the `gulp serve` cmd indirectly.

Comment (coder1033) [inline README.md]: Shouldn't we use just `npm run serve`?

File: README.md [hunk 2]
```diff
@@ -6,7 +6,8 @@ In order to run the project locally, you need few keys in order for 3rd party se
...
+4. Check route: http://localhost:5000/todos at postman
```
Comment (coder1033) [inline README.md]: Is this your base URI?

File: src/controllers/todo-controller.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  AppContext,
+  Errors,
+  ExtendedRequest,
+  ValidationFailure,
+} from '@typings';
+import {
+  createTodoValidator
+} from '@validators';
+
+export class TodoController extends BaseController {
```
Comment (coder1033) [inline src/controllers/todo-controller.ts]: Class name should be `TodoItemController`.

File: src/controllers/todo-controller.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  private initializeRoutes() {
+    this.router.post(
+      `${this.basePath}`,
+      createTodoValidator(),
+      this.createTodo,
+    );
+  }
+
+  private createTodo = async (
```
Comment (coder1033) [inline src/controllers/todo-controller.ts]: `createTodoItem`

File: src/controllers/todo-controller.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    if (failures.length > 0) {
+      const valError = new Errors.ValidationError(
+        res.__('DEFAULT_ERRORS.VALIDATION_FAILED'),
+        failures,
+      );
+      return next(valError);
+    }
+
+    const { title } = req.body;
+    const todo = await this.appContext.todoRepository.save(
```
Comment (coder1033) [inline src/controllers/todo-controller.ts]: `todoItem`

File: src/models/todo.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { BaseModel } from './base-model';
+import { LooseObject } from '@typings';
+
+export class TodoItem extends BaseModel {
```
Comment (coder1033) [inline src/models/todo.ts]: File name should be `todo-item`.

File: src/storage/mongoose/todo.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import { model, Schema } from "mongoose";
+
+const todoSchema: Schema = new Schema(
```
Comment (coder1033) [inline src/storage/mongoose/todo.ts]: `todoItemSchema`

File: src/validators/create-access-token-validator.ts [hunk 8]
```diff
@@ -9,7 +9,7 @@ const createAccessTokenValidator = (
     .isEmail()
     .withMessage('VALIDATION_ERRORS.INVALID_EMAIL')
     .custom(async (email) => {
```
Comment (coder1033) [inline src/validators/create-access-token-validator.ts]: Which email are you validating here?

File: src/validators/create-todo-validator.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import { check, ValidationChain } from 'express-validator';
+
+const createTodoValidator = (): ValidationChain[] => [
```
Comment (coder1033) [inline src/validators/create-todo-validator.ts]: `createTodoItemValidator`

File: test/mocks/in-memory-mongo-store.ts [hunk 10]
```diff
@@ -1,15 +1,14 @@
...
 
 export class InMemoryMongoStore extends Mongoose.MongoStore {
   constructor() {
     super();
   }
 
   public async connect(): Promise<M> {
-    const mongod = new MongoMemoryServer();
-    const uri = await mongod.getConnectionString();
+    const uri: string = config.get('mongoDb.uri');
```
Comment (coder1033) [inline test/mocks/in-memory-mongo-store.ts]: Use `mongod.getUri()`. We should use MongoMemoryServer instead of our database for testing.

**Comment by coder1033** [comment]
Please also check if the title is string or not.

**Comment by coder1033** [comment]
No manual testing?

File: README.md (hunk 1 above)
Comment (vinay-jtc) [reply README.md]: Yes, can use `npm run serve`

File: README.md (hunk 2 above)
Comment (vinay-jtc) [reply README.md]: this is base URI for /todos.

File: src/validators/create-access-token-validator.ts (hunk 8 above)
Comment (vinay-jtc) [reply src/validators/create-access-token-validator.ts]: This email was part of account handle. I will delete this file. It's not for use.

File: README.md (hunk 2 above)
Comment (coder1033) [reply README.md]: You should only write the base URI of this project.

File: test/mocks/in-memory-mongo-store.ts (hunk 10 above)
Comment (vinay-jtc) [reply test/mocks/in-memory-mongo-store.ts]: Didn't got this.

File: README.md (hunk 2 above)
Comment (vinay-jtc) [reply README.md]: okay

File: src/controllers/todo-controller.ts (hunk 3 above)
Comment (vinay-jtc) [reply src/controllers/todo-controller.ts]: updated

File: src/controllers/todo-controller.ts (hunk 4 above)
Comment (vinay-jtc) [reply src/controllers/todo-controller.ts]: updated

File: src/controllers/todo-controller.ts (hunk 5 above)
Comment (vinay-jtc) [reply src/controllers/todo-controller.ts]: updated

File: src/models/todo.ts (hunk 6 above)
Comment (vinay-jtc) [reply src/models/todo.ts]: updated

File: src/storage/mongoose/todo.ts (hunk 7 above)
Comment (vinay-jtc) [reply src/storage/mongoose/todo.ts]: updated

File: src/validators/create-todo-validator.ts (hunk 9 above)
Comment (vinay-jtc) [reply src/validators/create-todo-validator.ts]: updated

File: test/mocks/in-memory-mongo-store.ts (hunk 10 above)
Comment (vinay-jtc) [reply test/mocks/in-memory-mongo-store.ts]: updated

**Comment by vinay-jtc** [comment]
Checked manual testing and test added for title is string or not.

File: src/controllers/todo-controller.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+  AppContext,
+  Errors,
+  ExtendedRequest,
+  ValidationFailure,
+} from '@typings';
+import {
+  createTodoItemValidator
+} from '@validators';
+
+export class TodoItemController extends BaseController {
```
Comment (coder1033) [inline src/controllers/todo-controller.ts]: File name - `todo-item-controller.ts`

File: src/storage/mongoose/todo.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import { model, Schema } from "mongoose";
+
+const todoItemSchema: Schema = new Schema(
```
Comment (coder1033) [inline src/storage/mongoose/todo.ts]: File name - `todo-item.ts`

File: src/storage/mongoose/todo.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+  },
+  {
+    collection: "todos",
+    timestamps: {
+      createdAt: "createdAt",
+      updatedAt: "updatedAt",
+    },
+  }
+);
+const todoItem = model("Todo", todoItemSchema);
```
Comment (coder1033) [inline src/storage/mongoose/todo.ts]: `TodoItem` instead of `Todo`

File: src/validators/create-todo-validator.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+import { check, ValidationChain } from 'express-validator';
+
+const createTodoItemValidator = (): ValidationChain[] => [
```
Comment (coder1033) [inline src/validators/create-todo-validator.ts]: File name - `create-todo-item-validator`

**Review by coder1033 — APPROVED** [review]

File: src/validators/create-todo-validator.ts (hunk 14 above)
Comment (vinay-jtc) [reply src/validators/create-todo-validator.ts]: Updated

File: src/storage/mongoose/todo.ts (hunk 13 above)
Comment (vinay-jtc) [reply src/storage/mongoose/todo.ts]: Updated

File: src/storage/mongoose/todo.ts (hunk 12 above)
Comment (vinay-jtc) [reply src/storage/mongoose/todo.ts]: Updated

File: src/controllers/todo-controller.ts (hunk 11 above)
Comment (vinay-jtc) [reply src/controllers/todo-controller.ts]: Updated

**Pull request merged**